src/hookproxy/sha1.cpp

```diff
@@ -0,0 +1,252 @@
+                m_unprocessedBytes = SHA_BLOCK_LENGTH;
+            }
+
+            SHAEngine::Process_Block(msgblock, digest);
+            bytesremaining = 0;
+            m_unprocessedBytes = 0;
+        }
+
+        std::memset(msgblock + bytesremaining, 0, 56 - bytesremaining);
+        mblockp = &msgblock[56];
+        // pad total length of data into last 8 bytes
+        *reinterpret_cast<uint32_t *>(msgblock + 56) = htobe32(totalbytes >> 29);
+        *reinterpret_cast<uint32_t *>(msgblock + 60) = htobe32(totalbytes << 3);
+
+        m_unprocessedBytes = SHA_BLOCK_LENGTH;
+        SHAEngine::Process_Block(&msgblock[0], digest);
+        m_unprocessedBytes = 0;
+
+        // format the digest into the correct byte order, big endian
+        std::memcpy(&m_finalDigest, &digest, sizeof(m_finalDigest));
+        finalp = reinterpret_cast<uint32_t *>(&m_finalDigest.H0);
+
+        for (int i = 0; i < 5; ++i) {
+            finalp[i] = htobe32(finalp[i]);
+        }
+
+        m_computed = true;
+        std::memcpy(output, &m_finalDigest, SHA_DIGEST_LENGTH);
+    }
+
+    return 20;
+}
+
+/**
+ * @brief Processes a 64 byte block into the hash..
+ */
+void SHAEngine::Process_Block(const void *input, SHADigest &digest) const
+{
+    const uint8_t *data = static_cast<const uint8_t *>(input);
+    int t;
+    uint32_t a;
+    uint32_t b;
+    uint32_t c;
+    uint32_t d;
+    uint32_t e;
+    uint32_t K;
+    uint32_t f;
+    uint32_t W[80];
+
+    // Copy and expand the message block (Rounds 0 to 15?)
+    for (t = 0; t < 16; ++t) {
+        W[t] = (data[t * 4] << 24) + (data[t * 4 + 1] << 16) + (data[t * 4 + 2] << 8) + data[t * 4 + 3];
+    }
+
+    for (; t < 80; ++t) {
+        W[t] = __rotl32(W[t - 3] ^ W[t - 8] ^ W[t - 14] ^ W[t - 16], 1);
+    }
+
+    // Initialize registers with the previous intermediate value.
+    a = digest.H0;
+    b = digest.H1;
+    c = digest.H2;
+    d = digest.H3;
+    e = digest.H4;
+
+    // main loop
+    for (t = 0; t < 80; ++t) {
+        if (t < 20) {
+            K = 0x5A827999;
+            f = (b & c) | ((~b) & d);
+
+        } else if (t < 40) {
+            K = 0x6ED9EBA1;
+            f = b ^ c ^ d;
+
+        } else if (t < 60) {
+            K = 0x8F1BBCDC;
+            f = (b & c) | (b & d) | (c & d);
+
+        } else {
+            K = 0xCA62C1D6;
+            f = b ^ c ^ d;
+        }
+
+        uint32_t temp = __rotl32(a, 5) + f + e + W[t] + K;
+        e = d;
+        d = c;
+        c = __rotl32(b, 30);
+        b = a;
+        a = temp;
+    }
+
+    // Compute the current intermediate hash value.
+    digest.H0 += a;
+    digest.H1 += b;
+    digest.H2 += c;
+    digest.H3 += d;
+    digest.H4 += e;
+}
+
+/**
+ * @brief Formats the hash into a string for printing.
+ */
+void SHAEngine::Print(const void *buffer, char *output)
+{
+    for (int i = 0; i < 20; i++) {
+        std::sprintf(&output[2 * i], "%02x", *(static_cast<const uint8_t *>(buffer) + i));
```

Comment:
This looks dangerous. What if `output` buffer is not size 2*20? Then we corrupt memory here, no? Any chance to rewrite this to be safe to call always?